## Hands-on 4
Let's continue to use the dataset from the wine magazine used in lecture to practice data transformation, grouping, and sorting.

<img src="https://secure.static.meredith.com/crt/store/covers/magazines/nmo/9826_l.jpg">

In [92]:
csvurl = (
    "https://gist.githubusercontent.com/clairehq/"
    + "79acab35be50eaf1c383948ed3fd1129/raw/407a02139ae1e134992b90b4b2b8c329b3d73a6a/winemag-data-130k-v2.csv"
)
import pandas as pd

wine = pd.read_csv(csvurl)
wine.drop(wine.columns[0], axis="columns", inplace=True)
wine.head()

country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

#### Question 1: ####  
What is the mean of the points column?

In [93]:
# According to Jones (2020) on page 22, we have wide ranges of methods to
# produce descriptive statistics
wine["points"].mean()

88.43403716087269

The mean of the points column is computed to be 88.43

#### Question 2: ####  
How many countries are present in this dataset? (Only count each country once)

In [94]:
# Display participating countries in the country column.
wine["country"].unique()

array(['Italy', 'Portugal', 'US', 'Spain', 'France', 'Germany',
       'Argentina', 'Chile', 'Australia', 'Austria', 'South Africa',
       'New Zealand', 'Israel', 'Hungary', 'Greece', 'Romania', 'Mexico',
       'Canada', nan, 'Turkey', 'Czech Republic', 'Slovenia',
       'Luxembourg', 'Croatia', 'Georgia', 'Uruguay', 'England',
       'Lebanon', 'Serbia', 'Brazil', 'Moldova', 'Morocco', 'Peru',
       'India', 'Bulgaria', 'Cyprus', 'Armenia', 'Switzerland',
       'Bosnia and Herzegovina', 'Ukraine', 'Slovakia', 'Macedonia'],
      dtype=object)

In [95]:
# Pass the returned object from the `unique` function to python's builtin `len()` function
# to output the length of the array of distinct country names.
len(wine["country"].unique())

42

In [96]:
wine["country"].nunique()

41

In [97]:
wine["country"].nunique(dropna=False)

42

According to [Difference in len() and .nunique with relation to pandas dataframe](https://stackoverflow.com/questions/65826794/difference-in-len-and-nunique-with-relation-to-pandas-dataframe)
Using Python's builtin `len` function gives 42. Invoking `nunique` function on the "country" column outputs 41. The reason why the `len` function outputs 42 is because it counts the column header row. You need to pass `dropna=False` in `nunique()`

The methods `unique` outputs a series and details 42 rows. There are 42  countries.

#### Question 3: ####
How many times does each country appeared in this dataset? Show each country and the corresponding count (show counts in ascending order)

In [98]:
# According to Jones (2020) on page 46, the `value_counts` displays the occurrences for each country
# Invoke the `value_counts()` methods to display the number of occurrences for each country in
# wine dataframe.
wine["country"].value_counts()

country
US                        27177
France                    11174
Italy                     10005
Spain                      3409
Portugal                   2963
Chile                      2258
Argentina                  1907
Austria                    1635
Australia                  1177
Germany                    1051
South Africa                737
New Zealand                 733
Israel                      259
Greece                      242
Canada                      108
Bulgaria                     68
Romania                      67
Hungary                      61
Uruguay                      61
Croatia                      44
Turkey                       43
Slovenia                     39
Georgia                      37
England                      36
Mexico                       31
Brazil                       31
Moldova                      30
Lebanon                      20
Morocco                      11
Peru                          8
Serbia                        7


#### Question 4: ####
Create a variable `adjusted_price` containing the adjusted price which is the price subtracted by the average price. *This is called **"centering" transformation** - a method commonly used in the preprocessing step before applying various machine learning algorithms.*

In [99]:
# Talked to Hannah on Discord (I am Eggplant) to clarify variable = column
adjusted_price = wine["price"] - wine["price"].mean()
adjusted_price

0              NaN
1       -20.232932
2       -21.232932
3       -22.232932
4        29.767068
           ...    
65494     9.767068
65495   -13.232932
65496   -15.232932
65497    -4.232932
65498   -25.232932
Name: price, Length: 65499, dtype: float64

#### Question 5: ####
What is the title of the wine that has the highest points-to-price ratio in the dataset?

In [100]:
# According to https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html
# Using `loc` we can pass a "Conditional that returns a boolean Series with column labels specified"
# According to https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.idxmax.html
# Using `idxmax` we can return the index of the max value
wine.loc[(wine["points"] / wine["price"]).idxmax(), ["designation"]]

designation    NaN
Name: 64590, dtype: object

#### Question 6: ####
Create a series `flavor_counts` that contains two values: the number of wines that has the word "tart" in the `description` column and the number of wines that has the word "berries" in the `description` column. The index of the Series should be "Tart" and "Berries" for the corresponding values.

In [101]:
# According to https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html
# and According to https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sum.html
# We can search and count the number of occurrences
# Look for the word "tart", not case-sensitive, not a substring of another word.
# Same applies to the word "berries"

tart_count = wine["description"].str.contains("tart").sum()
berries_count = wine["description"].str.contains("berries").sum()
flavour_counts = pd.Series([tart_count, berries_count], index=["Tart", "Berries"])
flavour_counts

Tart       4172
Berries    3503
dtype: int64

#### Question 7: ####
Let's convert the points into simple star ratings. A score of 90 or higher counts as 3 stars, a score of at least 80 but less than 90 is 2 stars. Any other score is 1 star.

Also, any wines from France should automatically get 3 stars, regardless of points.

Add this new column `star_ratings` to the dataframe with the number of stars for each wine in the dataset. 

In [102]:
# Followed the example from the lecture by DeBaets (2024) to create a function and
# use the `map` function
def make_stars(points):
    if points >= 90:
        return "***"
    elif points >= 80 & points < 90:
        return "**"
    else:
        return "*"


wine["star_ratings"] = wine["points"].map(make_stars)

#### Question 8: ####
Who are the most common wine reviewers in the dataset? Create a Series whose index is the taster_twitter_handle category from the dataset, and whose values count how many reviews each person wrote.

In [103]:
wine.groupby(["taster_twitter_handle"]).size()

taster_twitter_handle
@AnneInVino          1769
@JoeCz               2605
@bkfiona               11
@gordone_cellars     2032
@kerinokeefe         5313
@laurbuzz             938
@mattkettmann        3035
@paulgwine           4851
@suskostrzewa         593
@vboone              4696
@vossroger          13045
@wawinereport        2358
@wineschach          7752
@winewchristina         4
@worldwineguys        465
dtype: int64

#### Question 9: ####
What combination of countries and varieties are most common? Create a Series whose index is a MultiIndexof {country, variety} pairs. For example, a pinot noir produced in the US should map to {"US", "Pinot Noir"}. Sort the values in the Series in descending order based on wine count.

In [104]:
# According to https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html
# and according to https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html
# We can use `grouby` function by passing a tuple.
# Use `sort_values` to sort the values in descending order
wine.groupby(["country", "variety"]).size().sort_values(ascending=False)

country  variety                 
US       Pinot Noir                  4918
         Cabernet Sauvignon          3649
         Chardonnay                  3412
France   Bordeaux-style Red Blend    2380
Italy    Red Blend                   1870
                                     ... 
         Torbato                        1
         Vespaiolo                      1
         Vespolina                      1
         Vitovska                       1
Uruguay  Tempranillo-Tannat             1
Length: 1304, dtype: int64

#### Question 10 #####
Create a Series whose index is reviewers and whose values is the average score given out by that reviewer. Hint: you will need the `taster_name` and `points` columns.

In [105]:
# https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html
wine.groupby(["taster_name"])["points"].mean()

taster_name
Alexander Peartree    86.014286
Anna Lee C. Iijima    88.380506
Anne Krebiehl MW      90.587903
Carrie Dykes          86.644444
Christina Pickard     89.500000
Fiona Adams           87.090909
Jeff Jenssen          88.273504
Jim Gordon            88.604331
Joe Czerwinski        88.519770
Kerin O’Keefe         88.827969
Lauren Buzzeo         87.831557
Matt Kettmann         90.021087
Michael Schachner     86.904541
Mike DeSimone         89.030303
Paul Gregutt          89.095032
Roger Voss            88.678957
Sean P. Sullivan      88.666243
Susan Kostrzewa       86.408094
Virginie Boone        89.229557
Name: points, dtype: float64

## References

DeBaets, H. (2024). Additional Pandas Problems.ipynb. Unknown Publisher
Jones, M. (2020). Biological data exploration with Python, pandas and seaborn. Independently published.